File: ObasAnalyzerCSharp/ObasAnalyzerCSharp/ObasAnalyzerCSharp/InterfaceAmbiguaAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;
using Utilerias.ObasAnalyzerCSharp;

namespace Analyzer1
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class InterfaceAmbiguaAnalyzer : DiagnosticAnalyzer
    {
        #region "Reglas"
        internal const string Category = "OBAS-SmellArquitectura-MVC";

        internal static readonly DiagnosticDescriptor Regla001InterfaceAmbigua = new DiagnosticDescriptor(
            "RA14001",
            "RA14-001: Interface Ambigua",
            "La interface {0} tiene demasiados métodos",
            Category,
            DiagnosticSeverity.Error,
            true);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Regla001InterfaceAmbigua); } }

        public override void Initialize(AnalysisContext context)
        {
            context.EnableConcurrentExecution();
            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
            // Registra el tipo de sintaxis que se analiza para cada regla
            context.RegisterSyntaxNodeAction(AnalyzeRegla001InterfaceAmbigua, SyntaxKind.InterfaceDeclaration);
        }
        #endregion

        #region "Analizadores"

        /// <summary>
        /// Valida el cumplimiento de la regla
        /// "RA14-001: Interface Ambigua"
        /// </summary>
        /// <param name="context"></param>
        private static void AnalyzeRegla001InterfaceAmbigua(SyntaxNodeAnalysisContext context)
        {
            var interfaceDeclaration = (InterfaceDeclarationSyntax)context.Node;

            // Cuenta la cantidad de métodos presentes
            int cantidadMetodos = 0;
            foreach (var miembro in interfaceDeclaration.Members)
            {
                if (miembro is MethodDeclarationSyntax)
                {
                    cantidadMetodos++;
                }
            }

            // Si sobrepasa el limite reporte el error
            if (cantidadMetodos > Constantes.limiteMetodosInterface)
            {
                var diagnostic = Diagnostic.Create(Regla001InterfaceAmbigua, interfaceDeclaration.Identifier.GetLocation(), interfaceDeclaration.Identifier.Text);
                context.ReportDiagnostic(diagnostic);
            }
        }
        #endregion
    }
}

File: ObasAnalyzerVB/ObasAnalyzerVB/ObasAnalyzerVB/InterfaceAmbiguaAnalyzer.vb
Imports ObasAnalyzerVB.Utilerias.ObasAnalyzer

<DiagnosticAnalyzer(LanguageNames.VisualBasic)>
Public Class InterfaceAmbiguaAnalyzer
    Inherits DiagnosticAnalyzer

#Region "Reglas"
    Friend Const Category As String = "OBAS-SmellArquitectura"

    Friend Shared ReadOnly Regla001InterfaceAmbigua As DiagnosticDescriptor = New DiagnosticDescriptor(
            "RA14001",
            "RA14-001: Interface Ambigua ",
            "La interface {0} tiene demasiados métodos.",
            Category,
            DiagnosticSeverity.Error,
            True)


    Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
        Get
            Return ImmutableArray.Create(Regla001InterfaceAmbigua)
        End Get
    End Property

    Public Overrides Sub Initialize(ByVal context As AnalysisContext)
        ' Registra el tipo de sintaxis que se analiza para cada regla
        context.RegisterSyntaxNodeAction(AddressOf AnalyzeRegla001InterfaceAmbigua, SyntaxKind.InterfaceBlock)
    End Sub
#End Region

#Region "Analizadores"

    ''' <summary>
    ''' Valida el cumplimiento de la regla
    ''' "RA14-001: Interface Ambigua"
    ''' </summary>
    ''' <param name="context"></param>
    Private Sub AnalyzeRegla001InterfaceAmbigua(ByVal context As SyntaxNodeAnalysisContext)
        Dim interfaceDeclaration = DirectCast(context.Node, InterfaceBlockSyntax)

        ' Cuenta la cantidad de métodos presentes
        Dim cantidadMetodos As Integer = 0
        For Each miembro In interfaceDeclaration.Members
            If TypeOf miembro Is MethodStatementSyntax Then
                cantidadMetodos += 1
            End If
        Next

        ' Si sobrepasa el límite, reporte el error
        If cantidadMetodos > Constantes.limiteMetodosInterface Then
            Dim diag = Diagnostic.Create(Regla001InterfaceAmbigua, interfaceDeclaration.InterfaceStatement.Identifier.GetLocation(), interfaceDeclaration.InterfaceStatement.Identifier.Text)
            context.ReportDiagnostic(diag)
        End If
    End Sub

#End Region


End Class

